### PR TITLE
91 merge stable version to main

### PR DESCRIPTION
**Pull request by ClementToni**
# This pull request includes all the new features add during the last sprint: 
- pulsar
- redis
- system
- master

We have now a system working with multi node/client, and a stateless master.

⬇️⬇️⬇️⬇️⬇️⬇️⬇️
This PR represents all the work we have done over the last two months. Please take the time to _**really test**_ each feature individually, review the code related to your expertise, and do not hesitate to ask questions, highlight coding style issues, or point out coding errors in the comments.
⬆️⬆️⬆️⬆️⬆️⬆️⬆️

# to test :
please take attention when testing the code, I probably forgot something.
Individual points to test: 
- pulsar:
    - ./automation/run --pulsar
- master:
    - ./automation/run --master
- redis:
    - ./automation/run --redis

- global test :
    - ./automations/run --test ./automations/test_suites/basicGameTest.yaml 

⚠️ I really encourage you tu download the [redis extension of VSCODE](https://marketplace.visualstudio.com/items?itemName=Redis.redis-for-vscode) to understand how redis and master work. (port 6379)
 

### Code Cleanup:
* remove kafka related file/folder
* Removed unused state handling and RPC registration code from the client source files (`CelteClient.cpp`, `StateConnected.cpp`, `StateConnecting.cpp`). [[1]](diffhunk://#diff-b47a7fd8c5aec60095721a68a85a0190f850b6defa29e5514807483908ffaa7bL1-L23) [[2]](diffhunk://#diff-96c5101cd88c5061c6ed96dd99634d3cb9cb9cfba78e1b981926b22432eb3ec4L1-L46) [[3]](diffhunk://#diff-8eb23ea6ea02920d832ae3f5b196e732f0969b16986be82fcfb95ed7b8480e86L1-L75)

### CODE REVIEW

File: redis/docker-compose.yml
Comment (EliotJanvier) [inline redis/docker-compose.yml:1]: open for discussion but could we create a docker compose that has both pulsar & redis to up them at the same time? (maybe through a third docker compose that references the other two)

File: system/runtime/include/Logger.hpp
Comment (EliotJanvier) [inline system/runtime/include/Logger.hpp:1]: we should open an issue to plan to incorporate redis to this logger

File: system/runtime/include/nlohmann/detail/exceptions.hpp
Comment (EliotJanvier) [inline system/runtime/include/nlohmann/detail/exceptions.hpp:1]: i want a cool header like this too!

**Review by EliotJanvier — CHANGES_REQUESTED** [review]
ok overall but some minor changes (mostly configurability)

File: master/src/Program.cs [hunk 1]
```diff
@@ -1,14 +1,25 @@
...
+﻿
+using System;
 using System.Threading;
-using Confluent.Kafka;
 
 class Program
 {
     static void Main(string[] args)
     {
+        Redis.RedisClient redisClient = Redis.RedisClient.GetInstance("localhost:6379");
```
[the remark this answers is not in the document]
Comment (ClementToni) [reply master/src/Program.cs:9]: yes, right

File: redis/docker-compose.yml
Comment (ClementToni) [reply redis/docker-compose.yml:1]: why not create a **--stack** (or other name) with: redis/pulsar inside a docker file ?

File: system/runtime/include/Logger.hpp
Comment (ClementToni) [reply system/runtime/include/Logger.hpp:1]: Who should be responsible for it ? 

@EliotJanvier
@Tumulus0 
@EwenBriand

Comment (ClementToni) [reply system/runtime/include/Logger.hpp:1]: Do not hesitate to ask me for an explanation of the redis stack

**Comment by ClementToni** [comment]
Screenshot 2024-12-09 at 6 14 48 PM
Still have a segfault since last update: 

##step to recreate the bug: 
- ./run --pulsar 
- ./run --redis
- ./run --master
- ./tests/server.debug
- ./tests/client.debug

Client connect, start to receive data, then server crash :/

File: master/configFile.yml [hunk 2]
```diff
@@ -1,5 +1,9 @@
-chunks: 1
+container_image: masterceltearm
 cpu: 1
+grapes:
+- LeChateauDuMechant
+- LeChateauDuGentil
+kafka_brokers: 192.168.0.19:80
```
Comment (EliotJanvier) [inline master/configFile.yml:6]: ip is still static, use env instead?

**Review by EliotJanvier — APPROVED** [review]

File: master/configFile.yml (hunk 2 above)
Comment (ClementToni) [reply master/configFile.yml:6]: Config file is supposed to be change by the developer, there is no need for an environment variable here.

File: system/runtime/include/Logger.hpp
Comment (EliotJanvier) [reply system/runtime/include/Logger.hpp:1]: @EwenBriand  then

**Pull request merged**